### PR TITLE
Controllers

### PR DESCRIPTION
**Pull request by ghost**
This PR is for creating controllers for the control panels.

Had to move the acp module **mode** name from the common language file to the `info_acp_` file.
Also added a logging example, when the settings were updated.
Took most comments from your pages and boardrules extensions.

For now just created the `admin_controller`.
Will also look to create the `mod_controller` and `user_controller`.
Or do we want to name these after their respective control panels, eg: `acp_controller`, `mcp_controller`, `ucp_controller`?

And for later on, do we want to include interfaces or is that a step beyond?

Also, for consistency it might be better to rename the `controller/main.php` to `controller/main_controller.php`

### CODE REVIEW

File: skeleton/controller/admin_controller.php.twig [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ */
+
+namespace {{ EXTENSION.vendor_name }}\{{ EXTENSION.extension_name }}\controller;
+
+/**
+ * {{ EXTENSION.extension_display_name }} admin controller.
+ */
+class admin_controller
```
Comment (VSEphpbb) [inline skeleton/controller/admin_controller.php.twig]: Let's call this `app_controller` so we keep this and the other module controllers all consistent and generic.

File: skeleton/config/services.yml.twig [hunk 2]
```diff
@@ -13,6 +13,18 @@ services:
             - '@template'
             - '@user'
 
+{% endif %}
+{% if COMPONENT.acp %}
+    {{ EXTENSION.vendor_name }}.{{ EXTENSION.extension_name }}.controller.admin:
```
Comment (VSEphpbb) [inline skeleton/config/services.yml.twig]: lets do `controller.acp` and `acp_controller`

**Comment by VSEphpbb** [comment]
Yes we can rename main controller to main_controller.

I'm not sure about interfaces. Those are not needed, per se, and I'd rather not inundate users with loads of files to tweak on out of the box, nor would I want us to imply how they choose to code (that is, it's our job to make sure they code within our guidelines, but interfaces are not a phpbb coding guideline requirement). If interfaces were to be involved, I would only think it shold be done so if we added a Component option checkbox for interfaces, which would generate interfaces for all the components the user has selected, so we give the user the option to choose them.

**Comment by mrgoldy** [comment]
Now I am not sure anymore, are we allowed to use `trigger_error()` in control panels or only in the acp? Should I use `$this->helper->message()` in the mcp/ucp.

Changes mcp's form action from `U_POST_ACTION` to `U_MCP_ACTION` for consistency.

Perhaps you can have a look at more descriptive function names and descriptions for the mcp/ucp controllers?

In the mcp controller I return the language string 'no user specified' upon submission as we never request the entered username (let alone do anything with it). But that way the 'structure' of the controller is atleast correct. [`$submit` > `$errors` > `trigger_error()`]

Anyways, I'm out for now. Big surprise, out celebrating NYE.
See you in the new year and a happy new year! 🎉 🎆

**Comment by VSEphpbb** [comment]
We should never really be using `trigger_error`. Problem is that's the only way to show errors to the user within the ACP. But outside ACP, avoid it, so yes use the helper message in UCP and MCP controllers.

**Pull request merged**